src/com/google/javascript/jscomp/CheckJSDoc.java

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright 2015 The Closure Compiler Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.javascript.jscomp;
+
+import com.google.javascript.jscomp.NodeTraversal.AbstractPostOrderCallback;
+import com.google.javascript.rhino.JSDocInfo;
+import com.google.javascript.rhino.Node;
+import com.google.javascript.rhino.Token;
+
+/**
+ * Checks for misplaced, misused or deprecated JSDoc annotations.
+ *
+ * @author chadkillingsworth@gmail.com (Chad Killingsworth)
+ */
+final class CheckJSDoc implements CompilerPass {
+
+  public static final DiagnosticType MISPLACED_ANNOTATION =
+      DiagnosticType.warning("JSC_MISPLACED_ANNOTATION",
+          "Misplaced {0} annotation.{1}");
+
+  public static final DiagnosticType ANNOTATION_DEPRECATED =
+      DiagnosticType.warning("JSC_ANNOTATION_DEPRECATED",
+          "The {0} annotation is deprecated.{1}");
+
+  private final AbstractCompiler compiler;
+
+  CheckJSDoc(AbstractCompiler compiler) {
+    this.compiler = compiler;
+  }
+
+  @Override
+  public void process(Node externs, Node root) {
+    NodeTraversal.traverse(compiler, externs, new ProcessNodes(true));
+    NodeTraversal.traverse(compiler, root, new ProcessNodes(false));
+  }
+
+  final class ProcessNodes extends AbstractPostOrderCallback {
+
+    private final boolean isExternsPass_;
+
+    ProcessNodes(boolean isExterns) {
+      isExternsPass_ = isExterns;
+    }
+
+    @Override
+    public void visit(NodeTraversal t, Node n, Node parent) {
+      validateTypeAnnotations(t, n);
+      validateFunctionJsDoc(t, n);
+      validateMsgJsDoc(t, n);
+      validateDeprecatedJsDoc(t, n);
+      validateNoCollapse(t, n);
+    }
+
+    /**
+     * Checks that deprecated annotations such as @expose are not present
+     */
+    private void validateDeprecatedJsDoc(NodeTraversal t, Node n) {
+      JSDocInfo info = n.getJSDocInfo();
+      if (info == null) {
+        return;
+      }
+      if (info.isExpose()) {
+        t.getCompiler().report(
+            t.makeError(n, ANNOTATION_DEPRECATED, "@expose",
+                " Use @nocollapse or @export instead."));
+      }
+    }
+
+    /**
+     * Checks that JSDoc intended for a function is actually attached to a
+     * function.
+     */
```

Comment:
It's copied over from validateFunctionJsDoc, but you forgot to edit it to talk about nocollapse